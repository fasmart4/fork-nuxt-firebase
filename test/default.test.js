const { resolve } = require('path')
const fs = require('fs-extra')
const { Nuxt, Builder } = require('nuxt')
const FirebaseModule = require('..')

jest.mock('firebase/app', () => ({
  apps: [
    new Proxy({}, {
      get (target, name) {
        return jest.fn(() => ({}))
      }
    })
  ],
  messaging: {
    isSupported: jest.fn()
  }
}))

describe('default', () => {
  let nuxt
  const buildDir = resolve(__dirname, '.nuxt-default')

  beforeAll(async () => {
    const config = {
      rootDir: resolve(__dirname, '..'),
      buildDir,
      srcDir: resolve(__dirname, 'fixture'),
      modules: [
        [FirebaseModule, {
          injectModule: false,
          config: {
            // REQUIRED: Official config for firebase.initializeApp(config):
            apiKey: '<apiKey>',
            authDomain: '<authDomain>',
            databaseURL: '<databaseURL>',
            projectId: '<projectId>',
            storageBucket: '<storageBucket>',
            messagingSenderId: '<messagingSenderId>',
            appId: '<appId>',
            measurementId: '<measurementId>'
          },
          services: {
            analytics: true,
            auth: true,
            firestore: true,
            functions: true,
            messaging: true,
            performance: true,
            database: true,
            remoteConfig: true,
            storage: true
          }
        }]
      ]
    }
    config.dev = false

    nuxt = new Nuxt(config)
    const BundleBuilder = { build: _ => _ }
    const builder = new Builder(nuxt, BundleBuilder)
    await builder.build()
  }, 60000)

  afterAll(async () => {
    await nuxt.close()
  })

  test('plugin contents', async () => {
    const content = await fs.readFile(resolve(buildDir, 'firebase/index.js'), { encoding: 'utf8' })

    expect(content).toContain('auth,')
    expect(content).toContain('messaging,')

    expect(content).not.toContain('Mess')

    expect(content).toMatchSnapshot()
  })

  test('exec plugin (server)', async () => {
    const Plugin = await import(resolve(buildDir, 'firebase/index.js')).then(m => m.default || m)
    const ctx = {}
    const inject = jest.fn()

    await Plugin(ctx, inject)
    expect(inject).toHaveBeenCalledTimes(1)
    expect(ctx.$fire).toBeDefined()
    expect(ctx.$fireModule).toBeUndefined()
  })

  test('exec plugin (client)', async () => {
    process.client = true
    const Plugin = await import(resolve(buildDir, 'firebase/index.js')).then(m => m.default || m)
    const ctx = {}
    const inject = jest.fn()

    await Plugin(ctx, inject)
    expect(inject).toHaveBeenCalledTimes(1)
    expect(ctx.$fire).toBeDefined()
    expect(ctx.$fireModule).toBeUndefined()
  })
})
