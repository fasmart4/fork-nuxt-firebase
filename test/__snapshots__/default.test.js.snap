// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`default plugin contents 1`] = `
"import createApp from './app.js'

import authService from './service.auth.js'
import databaseService from './service.database.js'
import firestoreService from './service.firestore.js'
import storageService from './service.storage.js'
import functionsService from './service.functions.js'
import messagingService from './service.messaging.js'
import performanceService from './service.performance.js'
import analyticsService from './service.analytics.js'
import remoteConfigService from './service.remoteConfig.js'

export default async (ctx, inject) => {
  // Runtime configs
  const appConfig = ctx.$config && ctx.$config.firebaseConfig || {\\"apiKey\\":\\"\\\\u003CapiKey\\\\u003E\\",\\"authDomain\\":\\"\\\\u003CauthDomain\\\\u003E\\",\\"databaseURL\\":\\"\\\\u003CdatabaseURL\\\\u003E\\",\\"projectId\\":\\"\\\\u003CprojectId\\\\u003E\\",\\"storageBucket\\":\\"\\\\u003CstorageBucket\\\\u003E\\",\\"messagingSenderId\\":\\"\\\\u003CmessagingSenderId\\\\u003E\\",\\"appId\\":\\"\\\\u003CappId\\\\u003E\\",\\"measurementId\\":\\"\\\\u003CmeasurementId\\\\u003E\\"}

  const { firebase, session } = await createApp(appConfig, ctx)

  let servicePromises = []

  if (process.server) {
    servicePromises = [
      authService(session, firebase, ctx, inject),
    databaseService(session, firebase, ctx, inject),
    firestoreService(session, firebase, ctx, inject),
    storageService(session, firebase, ctx, inject),
    functionsService(session, firebase, ctx, inject),

    ]
  }

  if (process.client) {
    servicePromises = [
      authService(session, firebase, ctx, inject),
      databaseService(session, firebase, ctx, inject),
      firestoreService(session, firebase, ctx, inject),
      storageService(session, firebase, ctx, inject),
      functionsService(session, firebase, ctx, inject),
      messagingService(session, firebase, ctx, inject),
      performanceService(session, firebase, ctx, inject),
      analyticsService(session, firebase, ctx, inject),
      remoteConfigService(session, firebase, ctx, inject),

    ]
  }

  const [
    auth,
    database,
    firestore,
    storage,
    functions,
    messaging,
    performance,
    analytics,
    remoteConfig
  ] = await Promise.all(servicePromises)

  const fire = {
    auth: auth,
    database: database,
    firestore: firestore,
    storage: storage,
    functions: functions,
    messaging: messaging,
    performance: performance,
    analytics: analytics,
    remoteConfig: remoteConfig
  }

  inject('fire', fire)
  ctx.$fire = fire
}

function forceInject (ctx, inject, key, value) {
  inject(key, value)
  const injectKey = '$' + key
  ctx[injectKey] = value
  if (typeof window !== \\"undefined\\" && window.$nuxt) {
  // If clause makes sure it's only run when ready() is called in a component, not in a plugin.
    window.$nuxt.$options[injectKey] = value
  }
}"
`;
